src/java/test/org/apache/zookeeper/server/quorum/ReconfigRollingRestartCompatibilityTest.java

```diff
@@ -0,0 +1,275 @@
+        }
+
+        for (int i = 0; i < serverCount; ++i) {
+            Assert.assertTrue("waiting for server " + i + " being up",
+                    ClientBase.waitForServerUp("127.0.0.1:" + clientPorts.get(i),
+                            CONNECTION_TIMEOUT));
+        }
+
+        for (int i = 0; i < serverCount; ++i) {
+            verifyQuorum(i);
+            verifyConfig(mt[i]);
+        }
+
+        Map<Integer, String> oldServerAddress = new HashMap<>(serverAddress);
+        config = updateExistingQuorumConfig(Arrays.asList(3, 4), new ArrayList<Integer>());
+        serverCount = serverAddress.size();
+        Assert.assertEquals("Server count should be 5 after config update.", serverCount, 5);
+
+        mt = Arrays.copyOf(mt, mt.length + 2);
+        for (int i = 3; i < 5; ++i) {
+            mt[i] = new QuorumPeerTestBase.MainThread(i, clientPorts.get(i),
+                    config, false);
+            mt[i].start();
+            Assert.assertTrue("waiting for server " + i + " being up",
+                    ClientBase.waitForServerUp("127.0.0.1:" + clientPorts.get(i),
+                            CONNECTION_TIMEOUT));
+            verifyConfig(mt[i]);
+            verifyQuorum(i);
+        }
+
+        Set<String> expectedConfigs = new HashSet<>();
+        for (String conf : oldServerAddress.values()) {
+            // Remove "server.x=" prefix.
+            expectedConfigs.add(conf.substring(9));
+        }
+
+        for (int i = 0; i < 3; ++i) {
+            verifyConfig(mt[i], expectedConfigs);
+        }
+
+        for (int i = 0; i < serverCount; ++i) {
+            mt[i].shutdown();
+        }
+    }
+
+    // A successful write operation indicates a quorum is up.
+    private void verifyQuorum(int sid)
+            throws IOException, InterruptedException, KeeperException {
+        ZooKeeper zk = new ZooKeeper("127.0.0.1:" + clientPorts.get(sid),
+                ClientBase.CONNECTION_TIMEOUT, new Watcher() {
+            public void process(WatchedEvent event) {
+            }});
+
+        for (int j = 0; j < 30; j++) {
+            try {
+                zk.create("/" + Integer.toString(sid) + "." + Math.random(),
+                        "foobar".getBytes(),
+                        ZooDefs.Ids.OPEN_ACL_UNSAFE, CreateMode.PERSISTENT);
+                break;
+            } catch (KeeperException.ConnectionLossException e) {
+                if (j < 29) {
+                    Thread.sleep(1000);
+                } else {
+                    Assert.fail("client could not connect to reestablished quorum: " +
+                            "giving up after 30+ seconds.");
+                }
+            }
+        }
+    }
+
+    private void verifyConfig(QuorumPeerTestBase.MainThread mt) {
+        Set<String> expectedConfigs = new HashSet<>();
+        for (String config : serverAddress.values()) {
+            // Remove "server.x=" prefix.
+            expectedConfigs.add(config.substring(9));
```

Comment:
this seems fragile (works only for single digit ids, perhaps use the index of "="